beacon_node/store/src/hot_cold_store.rs

```diff
@@ -3361,20 +3378,106 @@ fn no_state_root_iter() -> Option<std::iter::Empty<Result<(Hash256, Slot), Error
+        // - `state.slot() - 1`: TODO if we don't immediately commit all each state to the DB
+        //   individually, we may be attempting to read a state summary that is stored in a DB ops
+        //   vector but not yet written to the DB. Also starting from this slot is wasteful as we
+        //   know that the target slot is `< state.slot() - SlotsPerHistoricalRoot`.
+        // - `state.slot() - SlotsPerHistoricalRoot`: The most efficient slot to start. But we risk
+        //   jumping to a state summary that has already been pruned. See the `max(.., split_slot)`
+        //   below
+        let oldest_slot_in_state_roots = from_state
+            .slot()
+            .saturating_sub(Slot::new(E::SlotsPerHistoricalRoot::to_u64()));
+
+        // Don't start with a slot that prior to the finalized state slot. We may be attempting to read
+        // a hot state summary that has already been pruned as part of the migration and error. HDiffs
+        // can reference diffs with a slot prior to the finalized checkpoint. But those are sparse so
+        // the probabiliy of hitting `MissingSummary` error is high. Instead, the summary for the
+        // finalized state is always available.
+        let start_slot = std::cmp::max(oldest_slot_in_state_roots, store.get_split_slot());
+
+        *from_state
+            .get_state_root(start_slot)
+            .map_err(|_| StateSummaryIteratorError::OutOfBoundsInitialSlot)?
+    };
+
+    let mut previous_slot = None;
+
+    loop {
+        let state_summary = store
+            .load_hot_state_summary(&state_root)
+            .map_err(|e| StateSummaryIteratorError::LoadSummaryError(Box::new(e)))?
+            .ok_or(StateSummaryIteratorError::MissingSummary(state_root))?;
+
+        // Protect against infinite loops if the state summaries are not strictly descending
+        if let Some(previous_slot) = previous_slot {
+            if state_summary.slot >= previous_slot {
+                return Err(StateSummaryIteratorError::CircularSummaries {
+                    state_root,
+                    state_slot: state_summary.slot,
+                    previous_slot,
+                });
+            }
+        }
+        previous_slot = Some(state_summary.slot);
+
+        match state_summary.slot.cmp(&target_slot) {
+            Ordering::Less => {
+                return Err(StateSummaryIteratorError::BelowTarget(state_summary.slot))
+            }
+            Ordering::Equal => return Ok(state_root),
+            Ordering::Greater => {} // keep going
+        }
+
+        // Jump to an older state summary that is ancestor of `state_root`
+        if target_slot <= state_summary.diff_base_state_root.slot {
+            // As an optimization use the HDiff state root to jump states faster
+            state_root = state_summary.diff_base_state_root.state_root;
+        } else {
+            // Else jump slot by slot
+            state_root = state_summary.previous_state_root;
```

Comment:
This could be 0x0 in the case of the split state itself due to the initialisation in the migration here:

https://github.com/sigp/lighthouse/blob/06b19e33815c55ca3bdf44893fd7105a3c2f55c9/beacon_node/beacon_chain/src/schema_change/migration_schema_v24.rs#L170-L180

I think this could be a problem when we need to get a diff base state root prior to the split?

Reply:
I think actually that `Hash256::ZERO` is unreachable: https://github.com/sigp/lighthouse/pull/6750/files#r2025919484

Reply:
Fixed in:

- https://github.com/sigp/lighthouse/pull/6750/commits/088cfa03deec97ca56eb6bea4adaae35a2fa37a6
- https://github.com/sigp/lighthouse/pull/6750/commits/1184d7237e7c3cef031d208a763a5b3b076df23b